src/qibo/models.py

```diff
@@ -198,70 +199,182 @@ def loss(params):
+
+    Args:
+        hamiltonian (:class:`qibo.base.hamiltonians.Hamiltonian`): problem Hamiltonian
+            whose ground state is sought.
+        mixer (:class:`qibo.base.hamiltonians.Hamiltonian`): mixer Hamiltonian.
+            If ``None``, :class:`qibo.hamiltonians.X` is used.
+        solver (str): solver used to apply the exponential operators.
+        callbacks (list): List of callbacks to calculate during evolution.
+        accelerators (dict): Dictionary of devices to use for distributed
+            execution. See :class:`qibo.tensorflow.distcircuit.TensorflowDistributedCircuit`
+            for more details. This option is available only when ``hamiltonian``
+            is a :class:`qibo.base.hamiltonians.TrotterHamiltonian`.
+        memory_device (str): Name of device where the full state will be saved.
+            Relevant only for distributed execution (when ``accelerators`` is
+            given).
+
+    Example:
+        ::
+
+            import numpy as np
+            from qibo import models, hamiltonians
+            # create XXZ Hamiltonian for four qubits
+            hamiltonian = hamiltonians.XXZ(4)
+            # create QAOA model for this Hamiltonian
+            qaoa = models.QAOA(hamiltonian)
+            # optimize using random initial variational parameters
+            # and default options and initial state
+            initial_parameters = 0.01 * np.random.random(4)
+            best_energy, final_parameters = qaoa.minimize(initial_parameters, method="BFGS")
+    """
+    from qibo import hamiltonians, optimizers
+    from qibo.config import K, DTYPES, log
+    from qibo.callbacks import Norm
+    from qibo.base.hamiltonians import HAMILTONIAN_TYPES
+
+    def __init__(self, hamiltonian, mixer=None, solver="exp", callbacks=[],
+                 accelerators=None, memory_device="/CPU:0"):
+        # list of QAOA variational parameters (angles)
+        self.params = None
         # problem hamiltonian
+        if not isinstance(hamiltonian, self.HAMILTONIAN_TYPES):
+            raise_error(TypeError, "Invalid Hamiltonian type {}."
+                                   "".format(type(hamiltonian)))
         self.hamiltonian = hamiltonian
+        self.nqubits = hamiltonian.nqubits
         # mixer hamiltonian (default = -sum(sigma_x))
-        if mixer == None:
-            self.mixer = self.hamiltonians.X(nqubits)
+        if mixer is None:
+            trotter = isinstance(
+                self.hamiltonian, self.hamiltonians.TrotterHamiltonian)
+            self.mixer = self.hamiltonians.X(self.nqubits, trotter=trotter)
         else:
+            if type(mixer) != type(hamiltonian):
+                  raise_error(TypeError, "Given Hamiltonian is of type {} "
+                                         "while mixer is of type {}."
+                                         "".format(type(hamiltonian),
+                                                   type(mixer)))
```

Comment:
Maybe it is simpler to just check that mixer is of the right type? (because if hamiltonian is right, but not mixer, this message will appear and the user might not know which of the two is wrong)

Reply:
Thanks for the review.

The reason I check it this way is that the hamiltonians may be either `Hamiltonian` or `TrotterHamiltonian`. Both types are correct as long as both `hamiltonian` and `mixer` are of the same type (any of the two). I believe there is no confusion because I already added a check that the `hamiltonian` is of proper type earlier (following Stefano's suggestion).

I fixed the rest points you suggested.

Reply:
>The reason I check it this way is that the hamiltonians may be either Hamiltonian or TrotterHamiltonian. Both types are correct as long as both hamiltonian and mixer are of the same type (any of the two). I believe there is no confusion because I already added a check that the hamiltonian is of proper type earlier (following Stefano's suggestion).

I see. Thanks!